### PR TITLE
virt-viewer: Add version 10.0-1.0

### LINKED ISSUE #5001: [Request] Add virt-viewer
Virt-viewer is an open source tool to connect to virtual machines. It can display VNC or SPICE, and uses libvirt to lookup the graphical connection details. It's release page provides windows msi packages: https://virt-manager.org/download/

### PR DESCRIPTION
**Pull request by gitercn**
- Closes #5001 

Virt-viewer is an open source tool to connect to virtual machines. It can connect to VNC or SPICE, and uses libvirt to lookup the graphical connection details. It's release page provides windows msi packages: https://virt-manager.org/download/

I had made an issue request: https://github.com/lukesampson/scoop-extras/issues/5001 then I write the manifest myself.

### CODE REVIEW

**Comment by gitercn** [comment]
Hi @Ash258  I noticed the AppVeyor build failed is caused by "First-time contributors need a maintainer to approve running workflows." Could any maintainer review and if there is no problem then merge it?

**Comment by Ash258** [comment]
Why do you lie?

image

**Comment by gitercn** [comment]
Oh! I just noticed the build failed reason after you have posted the picture. I did not lie. The github message confused me, it says "first-time contributors need a maintainer to approve running workflows" right above the build failed message. This lead me to believe the build failed reason is "first-time contributors need a maintainer to approve running workflows". I didn't checked the details. Only after you posted the picture I realized I need to click the details to see the reason caused the build failure.

I have fixed the json files issues and now it build ok.

github-said-first-time-contributors-need-a-maintainer-to-approve-running-workflow

File: bucket/virt-viewer.json [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+        [
+            "bin\\remote-viewer.exe",
+            "Remote Viewer"
+        ]
+    ],
+    "checkver": {
+        "url": "https://virt-manager.org/download/",
+        "regex": "virt-viewer-x64-([\\d.-]+)\\.msi"
+    },
+    "autoupdate": {
```
Comment (Ash258) [inline bucket/virt-viewer.json:31]: Autoupdate does not make sense until extract dir is changed

File: bucket/virt-viewer.json [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+        "32bit": {
+            "url": "https://virt-manager.org/download/sources/virt-viewer/virt-viewer-x86-10.0-1.0.msi",
+            "hash": "ada7d13136f8ec2cf574071aa0bc1c1b8cf14ee62614b0728cdce5a46b080934"
+        }
+    },
+    "extract_dir": "VirtViewer v10.0-256",
+    "bin": [
+        "bin\\virt-viewer.exe",
+        "bin\\remote-viewer.exe"
+    ],
```
Comment (Ash258) [inline bucket/virt-viewer.json:20]: what are available command line arguments?

File: bucket/virt-viewer.json [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+        "64bit": {
+            "url": "https://virt-manager.org/download/sources/virt-viewer/virt-viewer-x64-10.0-1.0.msi",
+            "hash": "f50c13602060184b3e2db5f5822ff1860ac2eec39660578774008892e3516166"
+        },
+        "32bit": {
+            "url": "https://virt-manager.org/download/sources/virt-viewer/virt-viewer-x86-10.0-1.0.msi",
+            "hash": "ada7d13136f8ec2cf574071aa0bc1c1b8cf14ee62614b0728cdce5a46b080934"
+        }
+    },
+    "extract_dir": "VirtViewer v10.0-256",
```
Comment (Ash258) [inline bucket/virt-viewer.json:16]: What is the `256`??

File: bucket/virt-viewer.json [hunk 4]
```diff
@@ -0,0 +1,41 @@
+{
+    "version": "10.0-1.0",
+    "description": "Open source tool to connect to virtual machines to display VNC or SPICE",
+    "homepage": "https://virt-manager.org/download/",
+    "license": "GPL v3+",
```
Comment (Ash258) [inline bucket/virt-viewer.json:5]: Invalid SPDX identifier

File: bucket/virt-viewer.json [hunk 5]
```diff
@@ -0,0 +1,41 @@
+{
+    "version": "10.0-1.0",
+    "description": "Open source tool to connect to virtual machines to display VNC or SPICE",
+    "homepage": "https://virt-manager.org/download/",
```
Comment (Ash258) [inline bucket/virt-viewer.json:4]: Why download/??

File: bucket/virt-viewer.json [hunk 6]
```diff
@@ -0,0 +1,41 @@
+{
+    "version": "10.0-1.0",
+    "description": "Open source tool to connect to virtual machines to display VNC or SPICE",
```
Comment (Ash258) [inline bucket/virt-viewer.json:3]:
```suggestion
    "description": "The virtual machine console viewer",
```

**Review by Ash258 — CHANGES_REQUESTED** [review]

File: bucket/virt-viewer.json (hunk 5 above)
Comment (gitercn) [reply bucket/virt-viewer.json:4]: I was not sure what to put here. I will change it to https://virt-manager.org

File: bucket/virt-viewer.json (hunk 4 above)
Comment (gitercn) [reply bucket/virt-viewer.json:5]: I will change it to SPDX format "GPL-3.0-or-later"

File: bucket/virt-viewer.json (hunk 3 above)
Comment (gitercn) [reply bucket/virt-viewer.json:16]: This number is found when I install the virt-viewer-x86-10.0-1.0.msi all the files are installed to C:\Program Files\VirtViewer v10.0-256. Also when I extract it using lessmsi, all the files are under the VirtViewer v10.0-256 directory. I don't know what is the `256` meaning.  After searched their code at https://gitlab.com/virt-viewer/virt-viewer, I still don't know where the `256` comes from.

File: bucket/virt-viewer.json (hunk 2 above)
Comment (gitercn) [reply bucket/virt-viewer.json:20]: Do you mean the options of the two exe programs? I'm not sure how to write this section.

```
  virt-viewer.exe [OPTION?] -- ID|UUID|DOMAIN-NAME

Virtual machine graphical console

Help Options:
  -h, --help                                                           Show help options
  --help-all                                                           Show all help options
  --help-gtk                                                           Show GTK+ Options
  --help-gtk-vnc                                                       Show GTK-VNC Options
  --help-spice                                                         Show Spice Options

Application Options:
  -V, --version                                                        Display version information
  -z, --zoom=ZOOM                                                      Zoom level of window, in percentage
  -f, --full-screen                                                    Open in full screen mode (adjusts guest resolution to fit the client)
  -H, --hotkeys                                                        Customise hotkeys
  -r, --auto-resize=<never|always>                                     Automatically resize remote framebuffer
  -K, --keymap                                                         Remap keys format key=keymod+key e.g. F1=SHIFT+CTRL+F1,1=SHIFT+F1,ALT_L=Void
  --cursor=MODE                                                        Cursor display mode: 'local' or 'auto'
  -k, --kiosk                                                          Enable kiosk mode
  --kiosk-quit=<never|on-disconnect>                                   Quit on given condition in kiosk mode
  -v, --verbose                                                        Display verbose information
  --debug                                                              Display debugging information
  -d, --direct                                                         Direct connection with no automatic tunnels
  -a, --attach                                                         Attach to the local display using libvirt
  -c, --connect=URI                                                    Connect to hypervisor
  -w, --wait                                                           Wait for domain to start
  --reconnect                                                          Reconnect to domain upon restart
  --domain-name                                                        Select the virtual machine only by its name
  --id                                                                 Select the virtual machine only by its ID
  --uuid                                                               Select the virtual machine only by its UUID
  -s, --shared                                                         Share client session
```

```
  remote-viewer.exe [OPTION?] URI|VV-FILE

Remote viewer client
  -h, --help                                                           Show help options
Help Options:                                                          Show all help options
  -h, --help                                                           Show help options
  --help-all                                                           Show all help options
  --help-ovirt                                                         Show oVirt Options
  --help-gtk                                                           Show GTK+ Options
  --help-gtk-vnc                                                       Show GTK-VNC Options
  --help-spice                                                         Show Spice Optionsformation
  -z, --zoom=ZOOM                                                      Zoom level of window, in percentage
Application Options:                                                   Open in full screen mode (adjusts guest resolution to fit the client)
  -V, --version                                                        Display version information
  -z, --zoom=ZOOM                                                      Zoom level of window, in percentageffer
  -f, --full-screen                                                    Open in full screen mode (adjusts guest resolution to fit the client)_L=Void
  -H, --hotkeys                                                        Customise hotkeysde: 'local' or 'auto'
  -r, --auto-resize=<never|always>                                     Automatically resize remote framebuffer
  -K, --keymap                                                         Remap keys format key=keymod+key e.g. F1=SHIFT+CTRL+F1,1=SHIFT+F1,ALT_L=Void
  --cursor=MODE                                                        Cursor display mode: 'local' or 'auto'
  -k, --kiosk                                                          Enable kiosk mode information
  --kiosk-quit=<never|on-disconnect>                                   Quit on given condition in kiosk modeunnels
  -v, --verbose                                                        Display verbose information using libvirt
  --debug                                                              Display debugging information
  -t, --title                                                          Set window titleto start
  -s, --shared                                                         Share client sessionupon restart
  --domain-name                                                        Select the virtual machine only by its name
  --id                                                                 Select the virtual machine only by its ID
  --uuid                                                               Select the virtual machine only by its UUID
  -s, --shared                                                         Share client session
```

File: bucket/virt-viewer.json (hunk 1 above)
Comment (gitercn) [reply bucket/virt-viewer.json:31]: Is there a way to automatically the extract_dir of VirtViewer v10.0-*? Or should I just delete the autoupdate section?

File: bucket/virt-viewer.json (hunk 3 above)
Comment (gitercn) [reply bucket/virt-viewer.json:16]: They have replied: "This value 256 reflects the way we encode the buildid ("-1.0") part of the version number. Windows only permits a single value in this context, but we need two values, so we have to encode them. A future "-1.1" would be encoded to 257, and "-2.0" would become 512, and so on."

I have found a way to convert the 256 to 1.0 in order to update the extract_dir. So I have removed the deleted the autoupdate section in the manifest. If anyone have a way to convert it please help me edit it.

File: bucket/virt-viewer.json (hunk 1 above)
Comment (gitercn) [reply bucket/virt-viewer.json]: They have replied: "This value 256 reflects the way we encode the buildid ("-1.0") part of the version number. Windows only permits a single value in this context, but we need two values, so we have to encode them. A future "-1.1" would be encoded to 257, and "-2.0" would become 512, and so on."

I have found a way to convert the 256 to 1.0 in order to update the extract_dir. So I have removed the deleted the autoupdate section in the manifest. If anyone have a way to convert it please help me edit it.

**Review by gitercn — COMMENTED** [review]
I have removed the autoupdate section, could you please review again?

**Comment by rashil2000** [comment]
If you are unable to determine the directory to be extracted (through version string), you should move all the files to the root in a post_install script.

**Comment by gitercn** [comment]
Hi @rashil2000, do you mean I should move all the files to the `C:\Users\username\scoop\apps\virt-viewer` folder? I tried add a post_install script like this `"post_install": "Move-Item \"$original_dir\\*\" \"$original_dir\\..\""`. But the shortcut and bin still point to its current folder `C:\Users\username\scoop\apps\virt-viewer\current\bin\remote-viewer.exe`.

I also tried to search for examples https://github.com/ScoopInstaller/Extras/search?q=post_install+Move-Item. But I can't find cases like this. What to do? Is there any examples?

**Comment by rashil2000** [comment]
Sorry, I meant to say pre_install. I'll make the change.

**Review by rashil2000 — APPROVED** [review]

**Review by issaclin32 — APPROVED** [review]

**Pull request merged**